### PR TITLE
spirv-fuzz: add TransformationAddRelaxedDecoration

### LINKED ISSUE #3502: spirv-fuzz: transformation to add Relaxed to live safe functions and dead blocks

### PR DESCRIPTION
**Pull request by antnkarp**
Add TransformationAddRelaxedDecoration, which adds the RelaxedPrecision decoration to ids of numeric instructions (those yielding 32-bit ints or floats) in dead blocks.

Fixes #3502

### CODE REVIEW

File: test/fuzz/transformation_add_relaxed_decoration_test.cpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+  FactManager fact_manager;
+  spvtools::ValidatorOptions validator_options;
+  TransformationContext transformation_context(&fact_manager,
+                                               validator_options);
+
+  TransformationAddDeadBlock transformation1(100, 5, true);
+  ASSERT_TRUE(
+      transformation1.IsApplicable(context.get(), transformation_context));
+  transformation1.Apply(context.get(), &transformation_context);
```
Comment (paulthomson) [inline test/fuzz/transformation_add_relaxed_decoration_test.cpp:75]: I think we should probably avoid creating and applying transformations other than `TransformationAddRelaxedDecoration`. Consider `fact_manager.AddFactBlockIsDead`

File: test/fuzz/transformation_add_relaxed_decoration_test.cpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+      MakeInstructionDescriptor(100, SpvOpBranch, 0);
+
+  // SpvOpISub is numeric and is in the dead block, so transformation should be
+  // applicable.
+  auto transformation2 = TransformationEquationInstruction(
+      25, SpvOpISub, {9, 11}, return_instruction);
+  ASSERT_TRUE(
+      transformation2.IsApplicable(context.get(), transformation_context));
+  transformation2.Apply(context.get(), &transformation_context);
+  ASSERT_TRUE(IsValid(env, context.get()));
```
Comment (paulthomson) [inline test/fuzz/transformation_add_relaxed_decoration_test.cpp:87]: Same thing here and below. Do we need equation facts? You can just change the spirv assembly above if you want to add SpvOpISub etc.

Comment (paulthomson) [reply test/fuzz/transformation_add_relaxed_decoration_test.cpp:87]: Oh I see, you probably wanted to add stuff to the dead block. But by using the fact manager as explained above, you can mark any blocks as dead.

Comment (antnkarp) [reply test/fuzz/transformation_add_relaxed_decoration_test.cpp:87]: Thanks, I'll try this.

File: source/fuzz/fuzzer_pass_add_relaxed_decorations.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+FuzzerPassAddRelaxedDecorations::~FuzzerPassAddRelaxedDecorations() = default;
+
+void FuzzerPassAddRelaxedDecorations::Apply() {
+  // Consider every instruction in every block in every function.
+  for (auto& function : *GetIRContext()->module()) {
+    for (auto& block : function) {
+      for (auto& inst : block) {
+        // Randomly choose whether to apply the RelaxedPrecision decoration
+        // to this numeric instruction.
```
[the remark this answers is not in the document]
Comment (antnkarp) [reply source/fuzz/fuzzer_pass_add_relaxed_decorations.cpp:37]: Done: added check for IsNumeric

File: test/fuzz/transformation_add_relaxed_decoration_test.cpp (hunk 1 above)
Comment (antnkarp) [reply test/fuzz/transformation_add_relaxed_decoration_test.cpp]: Done. Thanks.

File: test/fuzz/transformation_add_relaxed_decoration_test.cpp (hunk 2 above)
Comment (antnkarp) [reply test/fuzz/transformation_add_relaxed_decoration_test.cpp]: Done

File: source/fuzz/fuzzer_pass_add_relaxed_decorations.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulthomson) [reply source/fuzz/fuzzer_pass_add_relaxed_decorations.cpp:37]: Sorry. I meant that the comment is misleading. I don’t think you need to check IsNumeric.

[the remark this answers is not in the document]
Comment (antnkarp) [reply source/fuzz/fuzzer_pass_add_relaxed_decorations.cpp]: Oh, thanks, I misunderstood. Fixed.

**Pull request merged**